lib/membrane/video_compositor/handler/callback_ctx.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Membrane.VideoCompositor.Handler.CallbackCtx do
+  @moduledoc """
+  Structure representing a common part of the context
+  for all of the callbacks.
+  """
+  alias Membrane.{Pad, Time}
+  alias Membrane.VideoCompositor.Scene
+
+  defmacro __using__(fields) do
+    default_fields =
+      quote do
+        [
+          input_pads: list(Pad.ref_t()),
+          scenes_queue: list({scene :: Scene.t(), start_scene_timestamp :: Time.t()}),
+          current_scene: Scene.t()
+        ]
+      end
+
+    quote do
+      @type t :: %__MODULE__{unquote_splicing(fields ++ default_fields)}
+
+      fields_names = unquote(Keyword.keys(fields))
+      default_fields_names = unquote(Keyword.keys(default_fields))
+
+      @enforce_keys Module.get_attribute(__MODULE__, :enforce_keys, fields_names) ++
+                      default_fields_names
+
+      defstruct fields_names ++ default_fields_names
```

Comment:
Here you are taking advantage that the quoted representation of the list is a list itself. And as far as I can see it won't allow to define custom default values:
```
defmodule X do
 use CallbackCtx, [x: 1, y: 2]
end

struct =.%X{input_pads: ..., scenes_queue: ..., current_scene: ...} # error
```

Reply:
Oh, apparently I have got it all wrong :D The "values" passed in the keyword list are about to be type specification, not the default values of the resulting keyword list